File: FBControlCore/Utility/FBArchiveOperations.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBArchiveOperations.h"

#import "FBControlCoreError.h"
#import "FBControlCoreLogger.h"
#import "FBProcess.h"
#import "FBProcessBuilder.h"

NSString *const BSDTarPath = @"/usr/bin/bsdtar";

@implementation FBArchiveOperations

+ (FBFuture<NSString *> *)extractArchiveAtPath:(NSString *)path toPath:(NSString *)extractPath overrideModificationTime:(BOOL)overrideMTime logger:(id<FBControlCoreLogger>)logger
{
  return [[[[[[[FBProcessBuilder
    withLaunchPath:BSDTarPath]
    withArguments:@[overrideMTime ? @"-zxpm" : @"-zxp", @"-C", extractPath, @"-f", path]]
    withStdErrToLoggerAndErrorMessage:logger.debug]
    withStdOutToLogger:logger.debug]
    withTaskLifecycleLoggingTo:logger]
    runUntilCompletionWithAcceptableExitCodes:[NSSet setWithObject:@0]]
    mapReplace:extractPath];
}

+ (NSArray<NSString *> *)commandToExtractFromStdInWithExtractPath:(NSString *)extractPath overrideModificationTime:(BOOL)overrideMTime compression:(FBCompressionFormat)compression
{
  NSArray<NSString *> *extractCommand = @[overrideMTime ? @"-zxpm" : @"-zxp", @"-C", extractPath, @"-f", @"-"];
  if (compression == FBCompressionFormatZSTD) {
    extractCommand = @[@"--use-compress-program", @"pzstd -d", overrideMTime ? @"-xpm" : @"-xp", @"-C", extractPath, @"-f", @"-"];
  }
  return extractCommand;
}

+ (FBFuture<NSString *> *)extractArchiveFromStream:(FBProcessInput *)stream toPath:(NSString *)extractPath overrideModificationTime:(BOOL)overrideMTime logger:(id<FBControlCoreLogger>)logger compression:(FBCompressionFormat)compression
{
  return [[[[[[[[FBProcessBuilder
    withLaunchPath:BSDTarPath]
    withArguments:[self commandToExtractFromStdInWithExtractPath:extractPath overrideModificationTime:overrideMTime compression:compression]]
    withStdIn:stream]
    withStdErrToLoggerAndErrorMessage:logger.debug]
    withStdOutToLogger:logger.debug]
    withTaskLifecycleLoggingTo:logger]
    runUntilCompletionWithAcceptableExitCodes:[NSSet setWithObject:@0]]
    mapReplace:extractPath];
}

+ (FBFuture<NSString *> *)extractGzipFromStream:(FBProcessInput *)stream toPath:(NSString *)extractPath logger:(id<FBControlCoreLogger>)logger
{
  return [[[[[[[[FBProcessBuilder
    withLaunchPath:@"/usr/bin/gunzip"]
    withArguments:@[@"--to-stdout"]]
    withStdIn:stream]
    withStdErrToLoggerAndErrorMessage:logger.debug]
    withStdOutPath:extractPath]
    withTaskLifecycleLoggingTo:logger]
    runUntilCompletionWithAcceptableExitCodes:[NSSet setWithObject:@0]]
    mapReplace:extractPath];
}

+ (FBFuture<FBProcess<NSNull *, NSInputStream *, id> *> *)createGzipForPath:(NSString *)path logger:(id<FBControlCoreLogger>)logger
{
  return (FBFuture<FBProcess<NSNull *, NSInputStream *, id> *> *) [[[[[[FBProcessBuilder
    withLaunchPath:@"/usr/bin/gzip"]
    withArguments:@[@"--to-stdout", path]]
    withStdErrToLoggerAndErrorMessage:logger]
    withStdOutToInputStream]
    withTaskLifecycleLoggingTo:logger]
    start];
}


+ (FBFuture<FBProcess<id, NSData *, id> *> *)createGzipDataFromProcessInput:(FBProcessInput *)input logger:(id<FBControlCoreLogger>)logger
{
  return (FBFuture<FBProcess<id, NSData *, id> *> *) [[[[[[[FBProcessBuilder
    withLaunchPath:@"/usr/bin/gzip"]
    withArguments:@[@"-", @"--to-stdout"]]
    withStdIn:input]
    withStdErrToLoggerAndErrorMessage:logger]
    withStdOutInMemoryAsData]
    withTaskLifecycleLoggingTo:logger]
    runUntilCompletionWithAcceptableExitCodes:[NSSet setWithObject:@0]
  ];
}


+ (FBFuture<FBProcess<NSNull *, NSInputStream *, id> *> *)createGzippedTarForPath:(NSString *)path logger:(id<FBControlCoreLogger>)logger
{
  NSError *error = nil;
  FBProcessBuilder<NSNull *, NSData *, id> *builder = [self createGzippedTarTaskBuilderForPath:path logger:logger error:&error];
  if (!builder) {
    return [FBFuture futureWithError:error];
  }
  return [[builder
    withStdOutToInputStream]
    start];
}

+ (FBFuture<NSData *> *)createGzippedTarDataForPath:(NSString *)path queue:(dispatch_queue_t)queue logger:(id<FBControlCoreLogger>)logger
{
  NSError *error = nil;
  FBProcessBuilder<NSNull *, NSData *, id> *builder = [self createGzippedTarTaskBuilderForPath:path logger:logger error:&error];
  if (!builder) {
    return [FBFuture futureWithError:error];
  }
  return [[builder
    runUntilCompletionWithAcceptableExitCodes:[NSSet setWithObject:@0]]
    onQueue:queue map:^(FBProcess<NSNull *, NSData *, id<FBControlCoreLogger>> *result) {
      return [result stdOut];
    }];
}

#pragma mark Private

+ (FBProcessBuilder<NSNull *, NSData *, id> *)createGzippedTarTaskBuilderForPath:(NSString *)path logger:(id<FBControlCoreLogger>)logger error:(NSError **)error
{
  BOOL isDirectory;
  if (![NSFileManager.defaultManager fileExistsAtPath:path isDirectory:&isDirectory]) {
    return [[FBControlCoreError
      describeFormat:@"Path for tarring %@ doesn't exist", path]
      fail:error];
  }

  NSString *directory;
  NSString *fileName;
  if (isDirectory) {
    directory = path;
    fileName = @".";
    [logger.info logFormat:@"%@ is a directory, tarring with it as the root.", directory];
    if ([[NSFileManager.defaultManager contentsOfDirectoryAtPath:path error:nil] count] < 1) {
      [logger.info logFormat:@"Attempting to tar directory at path %@, but it has no contents", path];
    }
  } else {
    directory = path.stringByDeletingLastPathComponent;
    fileName = path.lastPathComponent;
    [logger.info logFormat:@"%@ is a file, tarring relative to it's parent %@", path, directory];
    NSDictionary<NSString *, id> *fileAttributes = [NSFileManager.defaultManager attributesOfItemAtPath:path error:nil];
    NSUInteger fileSize = [fileAttributes[NSFileSize] unsignedIntegerValue];
    if (fileSize <= 0) {
      [logger.info logFormat:@"Attempting to tar file at path %@, but it has no content", path];
    }
  }

  return (FBProcessBuilder<NSNull *, NSData *, id> *) [[[[[FBProcessBuilder
    withLaunchPath:BSDTarPath]
    withArguments:@[@"-zvc", @"-f", @"-", @"-C", directory, fileName]]
    withStdOutInMemoryAsData]
    withStdErrToLoggerAndErrorMessage:logger]
    withTaskLifecycleLoggingTo:logger];
}

@end

File: FBControlCore/Utility/FBArchiveOperations.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public let BSDTarPath = "/usr/bin/bsdtar"

/// The compression types available.
public enum FBCompressionFormat: UInt {
  case GZIP = 1
  case ZSTD = 2
}

public enum ArchiveOperationsError: Error, LocalizedError {
  case pathDoesNotExist(path: String)

  public var errorDescription: String? {
    switch self {
    case let .pathDoesNotExist(path):
      return "Path for tarring \(path) doesn't exist"
    }
  }
}

/// Operations on zip/tar archives.
public enum FBArchiveOperations {

  /// Builds a command to extract from a file on disk.
  public static func commandToExtractArchive(
    atPath path: String,
    toPath extractPath: String,
    overrideModificationTime overrideMTime: Bool,
    debugLogging: Bool
  ) -> [String] {
    let flags = flagStringForExtraction(overrideModificationTime: overrideMTime, debugLogging: debugLogging)
    return [flags, "-C", extractPath, "-f", path]
  }

  /// Extracts a tar or zip file archive to a directory. The file can be an uncompressed tar, a
  /// gzipped tar, or a zip.
  public static func extractArchive(
    atPath path: String,
    toPath extractPath: String,
    overrideModificationTime overrideMTime: Bool,
    logger: any ControlCoreLogger
  ) -> FBFuture<NSString> {
    let arguments = commandToExtractArchive(
      atPath: path,
      toPath: extractPath,
      overrideModificationTime: overrideMTime,
      debugLogging: false)
    return FBProcessBuilder<NSNull, NSData, NSData>
      .withLaunchPath(BSDTarPath, arguments: arguments)
      .withStdErr(toLoggerAndErrorMessage: logger.debug())
      .withStdOut(to: logger.debug())
      .withTaskLifecycleLogging(to: logger)
      .runUntilCompletion(withAcceptableExitCodes: [0])
      .mapReplace(extractPath as NSString)
      .retyped()
  }

  /// Builds a command to extract via stdin.
  public static func commandToExtractFromStdIn(
    withExtractPath extractPath: String,
    overrideModificationTime overrideMTime: Bool,
    compression: FBCompressionFormat,
    debugLogging: Bool
  ) -> [String] {
    switch compression {
    case .ZSTD:
      return ["--use-compress-program", "pzstd -d", overrideMTime ? "-xpm" : "-xp", "-C", extractPath, "-f", "-"]
    case .GZIP:
      let flags = flagStringForExtraction(overrideModificationTime: overrideMTime, debugLogging: debugLogging)
      return [flags, "-C", extractPath, "-f", "-"]
    }
  }

  /// Extracts a tar or zip stream archive to a directory. The stream can be an uncompressed tar, a
  /// gzipped tar, a zstd compressed tar, or a zip.
  public static func extractArchive(
    fromStream stream: FBProcessInput<AnyObject>,
    toPath extractPath: String,
    overrideModificationTime overrideMTime: Bool,
    logger: any ControlCoreLogger,
    compression: FBCompressionFormat
  ) -> FBFuture<NSString> {
    let arguments = commandToExtractFromStdIn(
      withExtractPath: extractPath,
      overrideModificationTime: overrideMTime,
      compression: compression,
      debugLogging: false)
    return FBProcessBuilder<NSNull, NSData, NSData>
      .withLaunchPath(BSDTarPath, arguments: arguments)
      .withStdIn(stream)
      .withStdErr(toLoggerAndErrorMessage: logger.debug())
      .withStdOut(to: logger.debug())
      .withTaskLifecycleLogging(to: logger)
      .runUntilCompletion(withAcceptableExitCodes: [0])
      .mapReplace(extractPath as NSString)
      .retyped()
  }

  /// Extracts a gzip from a stream to a single file. A plain gzip wrapping a single file is
  /// preferred when there's only a single file to transfer.
  public static func extractGzip(
    fromStream stream: FBProcessInput<AnyObject>,
    toPath extractPath: String,
    logger: any ControlCoreLogger
  ) -> FBFuture<NSString> {
    FBProcessBuilder<NSNull, NSData, NSData>
      .withLaunchPath("/usr/bin/gunzip", arguments: ["--to-stdout"])
      .withStdIn(stream)
      .withStdErr(toLoggerAndErrorMessage: logger.debug())
      .withStdOutPath(extractPath)
      .withTaskLifecycleLogging(to: logger)
      .runUntilCompletion(withAcceptableExitCodes: [0])
      .mapReplace(extractPath as NSString)
      .retyped()
  }

  /// Creates a gzipped archive compressing the data provided.
  public static func createGzipData(
    from input: FBProcessInput<AnyObject>,
    logger: any ControlCoreLogger
  ) -> FBFuture<FBSubprocess<AnyObject, NSData, AnyObject>> {
    FBProcessBuilder<NSNull, NSData, NSData>
      .withLaunchPath("/usr/bin/gzip", arguments: ["-", "--to-stdout"])
      .withStdIn(input)
      .withStdErr(toLoggerAndErrorMessage: logger)
      .withStdOutInMemoryAsData()
      .withTaskLifecycleLogging(to: logger)
      .runUntilCompletion(withAcceptableExitCodes: [0])
      .retyped()
  }

  /// Creates a gzip archive, returning a task that has an input stream attached to stdout. Read the
  /// input stream to obtain all of the gzip output of the file. To confirm that the stream has been
  /// correctly written, the caller should check the exit code of the returned task upon completion.
  public static func createGzip(
    forPath path: String,
    logger: any ControlCoreLogger
  ) -> FBFuture<FBSubprocess<NSNull, InputStream, AnyObject>> {
    FBProcessBuilder<NSNull, NSData, NSData>
      .withLaunchPath("/usr/bin/gzip", arguments: ["--to-stdout", path])
      .withStdErr(toLoggerAndErrorMessage: logger)
      .withStdOutToInputStream()
      .withTaskLifecycleLogging(to: logger)
      .start()
      .retyped()
  }

  /// Creates a gzipped tar archive, returning a task that has an input stream attached to stdout.
  /// Read the input stream to obtain the gzipped tar output. To confirm that the stream has been
  /// correctly written, the caller should check the exit code of the returned task upon completion.
  public static func createGzippedTar(
    forPath path: String,
    logger: any ControlCoreLogger
  ) -> FBFuture<FBSubprocess<NSNull, InputStream, AnyObject>> {
    do {
      let arguments = try gzippedTarArguments(forPath: path, logger: logger)
      return FBProcessBuilder<NSNull, NSData, NSData>
        .withLaunchPath(BSDTarPath, arguments: arguments)
        .withStdOutToInputStream()
        .withStdErr(toLoggerAndErrorMessage: logger)
        .withTaskLifecycleLogging(to: logger)
        .start()
        .retyped()
    } catch {
      return FBFuture(error: error)
    }
  }

  /// Creates a gzipped tar archive, returning the data of the tar.
  public static func createGzippedTarData(
    forPath path: String,
    queue: DispatchQueue,
    logger: any ControlCoreLogger
  ) -> FBFuture<NSData> {
    do {
      let arguments = try gzippedTarArguments(forPath: path, logger: logger)
      return FBProcessBuilder<NSNull, NSData, NSData>
        .withLaunchPath(BSDTarPath, arguments: arguments)
        .withStdOutInMemoryAsData()
        .withStdErr(toLoggerAndErrorMessage: logger)
        .withTaskLifecycleLogging(to: logger)
        .runUntilCompletion(withAcceptableExitCodes: [0])
        .onQueue(queue, map: { subprocess in subprocess.stdOut ?? NSData() })
        .retyped()
    } catch {
      return FBFuture(error: error)
    }
  }

  private static func flagStringForExtraction(overrideModificationTime overrideMTime: Bool, debugLogging: Bool) -> String {
    var flags = ["z", "x", "p"]
    if overrideMTime {
      flags.append("m")
    }
    if debugLogging {
      flags.append("v")
    }
    return "-\(flags.joined())"
  }

  /// A directory is tarred with itself as the root; a file is tarred relative to its parent.
  private static func gzippedTarArguments(forPath path: String, logger: any ControlCoreLogger) throws -> [String] {
    var isDirectory: ObjCBool = false
    guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory) else {
      throw ArchiveOperationsError.pathDoesNotExist(path: path)
    }

    let directory: String
    let fileName: String
    if isDirectory.boolValue {
      directory = path
      fileName = "."
      logger.info().log("\(directory) is a directory, tarring with it as the root.")
      if (try? FileManager.default.contentsOfDirectory(atPath: path))?.isEmpty ?? true {
        logger.info().log("Attempting to tar directory at path \(path), but it has no contents")
      }
    } else {
      directory = (path as NSString).deletingLastPathComponent
      fileName = (path as NSString).lastPathComponent
      logger.info().log("\(path) is a file, tarring relative to it's parent \(directory)")
      let fileSize = (try? FileManager.default.attributesOfItem(atPath: path))?[.size] as? UInt ?? 0
      if fileSize <= 0 {
        logger.info().log("Attempting to tar file at path \(path), but it has no content")
      }
    }
    return ["-zvc", "-f", "-", "-C", directory, fileName]
  }
}
